ouroboros-consensus/src/unstable-consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Adversarial.hs

```diff
@@ -600,28 +634,98 @@ uniformAdversarialChain mbAsc recipe g0 = wrap $ C.createV $ do
-                -- schema
-                hScgWin = C.UnsafeContains (C.windowStart carWin) (C.Count s)
-                -- number of active slots in the first stability window after
-                -- the intersection in the honest schema
-                hCount = BV.countActivesInV S.notInverted (C.sliceV hScgWin vH)
-                -- first stability window after the intersection in the alternative
-                -- schema
-                aScgWin = C.UnsafeContains (C.Count 0) (C.Count s)
-                -- slice of the alternative schema for the first stability
-                -- window after the intersection
-                aScgMv  = C.sliceMV aScgWin mv
-                -- intended number of empty slots in the first stability window
-                -- after the intersection in the alternative schema
-                emptyCountTarget = C.toVar $ S.complementActive S.notInverted (C.windowSize aScgWin) hCount C.+ 1
-            void $ BV.fillInWindow
-                S.inverted
-                (BV.SomeDensityWindow emptyCountTarget (C.windowSize aScgWin))
-                g
-                aScgMv
+    -- | Ensure the density of the adversarial schema is less than the density
+    -- of the honest schema in the first stability window after the intersection
+    -- and in any prefix window that contains the first race to the k+1st block.
+    --
+    -- Ensuring lower density of the prefix windows is necessary to avoid chains like
+    --
+    -- > k: 3
+    -- > s: 9
+    -- > H: 0111100
+    -- > A: 1110011
+    --
+    -- where the honest chain wins the race to the k+1st block, might win the
+    -- density comparison if the chains are extended to include a full stability
+    -- window after the intersection, but loses the density comparison if the
+    -- chains aren't extended.
+    --
+    -- For the sake of shrinking test inputs, we also prevent the above scenario
+    -- to occur in any intersection, not just the intersections near the end of
+    -- the chains.
+    --
+    ensureLowerDensityInWindows
+      :: R.StatefulGen sg (ST s) => RI.Race adv -> sg -> C.MVector adv SlotE s S.S -> ST s ()
+    ensureLowerDensityInWindows (RI.Race (C.SomeWindow _ w0)) g mv = do
+        let
+          -- A window after the intersection as short as the shortest of the
+          -- stability window or the first race to the k+1st block.
+          w0' = C.truncateWin w0 (C.Count s)
+          hCount = C.toVar $ BV.countActivesInV S.notInverted (C.sliceV w0' vA)
+
+        aCount <- ensureLowerDensityInWindow w0' g mv hCount
+
+        void $ foldM
+          updateDensityOfMv
+          (hCount, aCount)
+          (stablePrefixWindowsContaining w0')
+
+      where
+        -- Yields all windows of the adversarial schema with proper prefix @w@
+        -- that are prefixes of the first stability window after the
+        -- intersection.
+        stablePrefixWindowsContaining w =
```

Comment:
Is this equivalent? Seems more direct in my opinion, if so.

```
stablePrefixWindowsContaining w =
    let start = C.windowStart w
        size  = C.getCount (C.windowSize w)
        end   = s `min` C.getCount (C.lengthMV mv)
    in
    [ C.UnsafeContains start (C.Count size') | size' <- [ size+1 .. end ] ]
```

Reply:
Looks good to me. Adopted in 4798d0c.